include/seqan3/alignment/matrix/detail/aligned_sequence_builder.hpp

```diff
@@ -0,0 +1,250 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \brief Provides seqan3::detail::aligned_sequence_builder.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <type_traits>
+#include <vector>
+
+#include <seqan3/alignment/aligned_sequence/aligned_sequence_concept.hpp>
+#include <seqan3/alignment/matrix/detail/matrix_coordinate.hpp>
+#include <seqan3/alignment/matrix/trace_directions.hpp>
+#include <seqan3/alphabet/gap/gapped.hpp>
+#include <seqan3/range/decorator/gap_decorator.hpp>
+#include <seqan3/range/view/slice.hpp>
+#include <seqan3/range/view/view_all.hpp>
+#include <seqan3/std/concepts>
+#include <seqan3/std/ranges>
+
+namespace seqan3::detail
+{
+
+/*!\brief Transforms the given range types into their aligned sequence format.
+ * \ingroup alignment_matrix
+ * \tparam fst_rng_t The type of the first range to be transformed to a aligned sequence type;
+ *                   must model std::ranges::ViewableRange.
+ * \tparam sec_rng_t The type of the second range to be transformed to a aligned sequence type;
+ *                   must model std::ranges::ViewableRange.
+ *
+ * \details
+ *
+ * First this trait refines the original range types with seqan3::view::slice. Then it transforms the given range type
+ * into a seqan3::gap_decorator if the expression is not ill-formed, otherwise the transformed aligned sequence type
```

Comment:
in which cases is the gap decorator ill-formed?

Reply:
Whenever you have non sized or non random access range. That clearly depends on the gap dectorator

Reply:
ugh we have to change that :unamused: